rfcs/0005-range-queries.md

```diff
@@ -0,0 +1,256 @@
+# SlateDB Range Scans
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Public API](#public-api)
+- [Implementation](#implementation)
+   * [Checkpoint References](#checkpoint-references)
+   * [Memory Usage](#memory-usage)
+- [Rejected Alternatives & Follow-Ups](#rejected-alternatives-follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Discussion
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+References:
+
+* https://github.com/slatedb/slatedb/issues/8
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/blob/main/rfcs/0004-checkpoints.md
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan<T>(
+        &self,
+        range: &T
```

Comment:
What's expected here when I want to grab an iterator and call `seek`? Is `(..)` the expected way?

Reply:
Yeah, good question. Currently we specify `&T` so I think we could pass `&..`. Perhaps it is a little nicer to pass range by value?

Reply:
Ya, I'm partial to range by value, but that's a small nit. =)

Reply:
Sounds good to me. This might end up having some implications for `SstIterator` which uses a reference for the start key, but probably we can just change that to use `Bytes`.

Reply:
Cool. If it ends up getting in the way, feel free to change as needed.